ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -1839,6 +1839,15 @@ boolean doPhase1(ASTNode ast, QB qb, Phase1Ctx ctx_1, PlannerContext plannerCtx)
             doPhase1GetDistinctFuncExprs(qbp.getAggregationExprsForClause(ctx_1.dest)));
         break;
 
+      case HiveParser.TOK_QUALIFY:
+        if (!HiveConf.getBoolVar(conf, HiveConf.ConfVars.HIVE_CBO_ENABLED)) {
```

Comment:
I think when we fall-back to the non-cbo path (more likely in nature) this flag will not be changed see around:
https://github.com/apache/hive/blob/35d4532b0c08f4f5fbb5dc897c4330cba434bc7c/ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java#L706

I think `set hive.cbo.fallback.strategy=ALWAYS` could also be used to check if that would happen